Contrib/mikerizzo email communication

mikejrizzo opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/45998

## Description
Added functionality to the content pack to allow for bi-directional email communication on any XSOAR incident - regardless of whether it was fetched from an email or any other source.  Users can add the new 'Email Threads' layout tab to their own Incident layouts and initiate outbound email communications to end users.  Replies to the original email are linked back to the related incident and can be displayed and further replied to using the "Email Threads" layout.

The 'Email Threads' layout allows the user to switch between existing email threads on the incident and view them and/or reply to them.  CC and BCC recipients can be added on to email threads at any time when replying to an existing email thread.

File attachments can be added to new email threads or replies to existing ones.  Files attached to replies from end users are retrieved and uploaded to the incident.

Construction of new email messages is based on incident fields.  This also allows for the automation of email construction and sending by other automation tasks if needed.  Once the required fields are set, custom automation can run the 'SendEmailReply' script to send the message created by automation.

The original functionality of the 'Email Communication' layout is preserved.  The only modification that affects the original functionality is that the ability to define a custom email signature has been included in the send_reply of the SendEmailReply script.  The optional custom signature can be set with a list named "XSOAR - Email Communication Signature".

## Screenshots
![PR Screenshot](https://user-images.githubusercontent.com/12971863/156059997-2b339709-cad9-44ee-abd4-6448fbe0211c.png)



## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [ ] Tests
- [ ] Documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=17717) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=17717) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=17717) it.</sub>

yuvalbenshalom commented:
@mikejrizzo please check circle: Run validation errors and fix what you can from printed list.
especially error type ST106.
If you have any questions, let's talk on slack.

mikejrizzo commented:
> @mikejrizzo please check circle: Run validation errors and fix what you can from printed list. especially error type ST106. If you have any questions, let's talk on slack.

@yuvalbenshalom I've addressed most of the issues raised by CircleCI, including ST106.  Let me know if there's anything else I need to fix before we can move forward.  Thanks!

yuvalbenshalom commented on Packs/EmailCommunication/Scripts/SummarizeEmailThreads/SummarizeEmailThreads.py:
```diff
@@ -0,0 +1,114 @@
+        for thread_entry in email_threads:
+            # Loop through thread entries.  'EmailCommsThreadNumber' is indexed from 0 so we can use it for list indexing
+            # Trim off thread ID code for readability
+            thread_number = thread_entry['EmailCommsThreadNumber']
```
if this is needed in n int form in from this place on, please do it here, rather than turning it to int many times later in the code.

mikejrizzo replied:
Done.

yuvalbenshalom commented on Packs/EmailCommunication/Scripts/SummarizeEmailThreads/SummarizeEmailThreads.py:
```diff
@@ -0,0 +1,114 @@
+            bcc_addresses = thread_entry['EmailBCC']
+            recipients = thread_entry['EmailTo']
+
+            if 0 <= int(thread_number) < len(thread_summary_md):
```
you are using the thread numbers as an index in the list, and append the table_row for the thread number at the end of the list. If I understand correctly, this code can only work if thread numbers are ordered always 0,1,2,3 and none of them will be missed. or else this indexing: `thread_summary_md[int(thread_number)]` will get the wrong item in the list. is that what you meant to do here?

mikejrizzo replied:
Done.

mikejrizzo replied:
New entries in the `EmailThreads` context key should always be created with the `EmailCommsThreadNumber` starting at zero, with each new email thread not related to an existing entry getting its own new number incrementing by one each time.

There is no function within this content pack that would ever remove entries from `EmailThreads`, however I can see how it might be possible for actions taken by other tasks to make changes to this context key.

I have updated this function to use a dictionary of dictionaries to provide reliable indexing, then convert that to a list of dictionaries for MD formatting.

yuvalbenshalom commented on Packs/EmailCommunication/Scripts/SummarizeEmailThreads/SummarizeEmailThreads.py:
```diff
@@ -0,0 +1,114 @@
+import demistomock as demisto  # noqa: F401
+from CommonServerPython import *  # noqa: F401
+
```
this script should have unit tests.

yuvalbenshalom reviewed (changes requested):
@mikejrizzo I just managed to review this completly.
This is a bg contribution.
Please see my comments. some of them are actionable and some are notes for a call that we will set up.(I referred it explicitly)

other notes:
1. please fix the lint errors in `ci/circleci: Run Unit Testing And Lint` build. check this build  when you will make more changes.
2. we will fix the validation errors in `ci/circleci: Run Validations` together in a call.

mikejrizzo commented on Packs/EmailCommunication/Scripts/DisplayEmailHtmlThread/DisplayEmailHtmlThread.py:
```diff
@@ -0,0 +1,130 @@
+            }
+            thread_items.append(thread_dict)
+
+    attachments = None  # incident.get('attachment', {})
```
This was a remnant from earlier work when attachment data was being stored on on a per-Incident level instead of a per-message level.  I have removed the unneeded references to this attachment field.

mikejrizzo commented on Packs/EmailCommunication/Scripts/DisplayEmailHtmlThread/DisplayEmailHtmlThread.py:
```diff
@@ -0,0 +1,130 @@
+        str. Email reply.
+    """
+
+    single_reply = f'<html><body><b>From:</b> {email_from}<br><b>To:</b> {email_to}<br><b>CC:</b> {email_cc}<br>' \
```
I don't have access to test environments of o365 or gmail, but I don't believe any of these changes are dependent on the mail service so long as they offer the **!send-mail** and **!reply-mail** commands.

I would appreciate any additional help testing on other email services.

mikejrizzo commented on Packs/EmailCommunication/Scripts/DisplayEmailHtmlThread/DisplayEmailHtmlThread.py:
```diff
@@ -0,0 +1,130 @@
+    if thread_exists:
+        # Append all messages together, with the most recent on top
+        for thread in reversed(thread_items):
+            email_reply = set_email_reply(thread['email_from'], thread['email_to'], thread['email_cc'],
+                                          thread['email_subject'], thread['email_html'], thread['email_time'],
+                                          attachments, thread['email_attachments'])
```
The code that sets these in the first place should always create them, but certainly it's a good idea to make it as error proof as possible.  I have changed these references to use .get() and return "None" if they're not present.

mikejrizzo commented on Packs/EmailCommunication/Scripts/SendEmailReply/SendEmailReply_test.py:
```diff
@@ -144,3 +159,105 @@ def test_get_email_cc(current_cc, additional_cc, excepted):
     assert result == excepted
+
+
+def test_create_thread_context(mocker):
```
Added test documentation:

> """Unit test
        Given:
        - all required function arguments are provided
        When:
        - creating new context entry to store email thread data
        Then
        - validate that function calls demisto.executeCommand() with all arguments and data needed to properly create
          the required context entry
    """

mikejrizzo commented on Packs/EmailCommunication/Scripts/PreprocessEmail/PreprocessEmail_test.py:
```diff
@@ -255,20 +353,49 @@ def test_main(mocker):
+    assert expected_return == demisto.results.call_args[0][0]
+
 
+def test_create_thread_context(mocker):
```
Added test documentation:

>     """Unit test
        Given:
        - all required function arguments are provided
        When:
        - creating new context entry to store email thread data
        Then
        - validate that function calls demisto.executeCommand() with all arguments and data needed to properly create
          the required context entry
    """

mikejrizzo commented on Packs/EmailCommunication/Scripts/PreprocessEmail/PreprocessEmail_test.py:
```diff
@@ -255,20 +353,49 @@ def test_main(mocker):
+    assert EMAIL_THREADS[0]['Contents']['context']['EmailThreads'][3] == call_args.args[1]['arguments']['value']
+
+
+def test_get_email_related_incident_id_1(mocker):
```
Renamed these to be more descriptive, though it does make the function names quite long.  Let me know if you have a more concise suggestion I should use.

test_get_email_related_incident_id_1 --> test_get_email_related_incident_id_email_in_fields
test_get_email_related_incident_id_2 --> test_get_email_related_incident_id_email_in_context

JasBeilin commented on Packs/EmailCommunication/Scripts/SendEmailReply/SendEmailReply.yml:
```diff
@@ -28,14 +33,20 @@ comment: |-
+tags: []
+scripttarget: 0
+runonce: false
+runas: DBotWeakRole
```
This is the weak role, which is ok, but I see use of `demisto.executeCommand('executeCommandAt')` and `demisto.executeCommand('setIncident')` (which by the way should use execute_command instead) in this pr so it might fail on permissions.. Are we sure it is working?

mikejrizzo replied:
I have replaced `executeCommandAt` in this script with `appendContext`.  I had copied this function from PreprocessEmail.py where `executeCommandAt` is indeed needed to set context on another existing incident.

Even with `executeCommandAt` this was working fine with DBotWeakRole, but after my change it won't be an issue either way.

mikejrizzo commented on Packs/EmailCommunication/Scripts/SendEmailReply/SendEmailReply_test.py:
```diff
@@ -144,3 +159,105 @@ def test_get_email_cc(current_cc, additional_cc, excepted):
+        Preparing email message to send
+    Then
+     - validate that logic is followed as expected
+     - for case with new_thread = 'false', validate proper email details are pulled from context
```
I've added a separate block of information for each permutation of the test function.  This makes the documentation block for this one function quite long, but there's enough re-used setup for this test that it makes sense to have a single function than have a separate one for each scenario.

>  Unit test scenario - new_thread = 'n/a'
>         Given
>          - Script is called to send an email message
>         When
>          - new_thread = 'n/a', indicating use by the generic 'Email Communication' layout workflow replying to message
>         Then
>          - validate that the new email is constructed and the send_reply function is called successfully
>     Unit test scenario - new_thread = 'true'
>         Given
>          - Script is called to send an email message
>         When
>          - new_thread = 'true', indicating use by incident types besides 'Email Communication' and sending a
>          first-contact outbound email message
>         Then
>          - validate that the new email is constructed and the send_new_email function is called successfully
>     Unit test scenario - new_thread = 'false'
>         Given
>          - Script is called to send an email message
>         When
>          - new_thread = 'false', indicating use by incident types besides 'Email Communication' and sending a
>          reply to an existing email chain
>          - 'EmailThreads' context key is type dict, indicating an attempt to reply to an email chain that has never
>          received a reply from the end user and we must send a new email message rather than reply to one
>         Then
>          - validate that the new email is constructed and resend_first_contact is called successfully
>     Unit test scenario - new_thread = 'false'
>         Given
>          - Script is called to send an email message
>         When
>          - new_thread = 'false', indicating use by incident types besides 'Email Communication' and sending a
>          reply to an existing email chain
>          - 'EmailThreads' context key is type list, indicating there is at least one reply to the first-contact email
>         Then
>          - validate that the new email is constructed, re-using correct items from the existing email thread entries
>          - validate that send_reply is called successfully
>          - validate that create_thread_context is successfully called

mikejrizzo commented on Packs/EmailCommunication/Scripts/SendEmailReply/SendEmailReply_test.py:
```diff
@@ -144,3 +159,105 @@ def test_get_email_cc(current_cc, additional_cc, excepted):
+def test_main(new_thread, thread_input_type, mocker):
+    """Unit test
+    Given
+     - new_thread = 'n/a'
+     - new_thread = 'true'
+     - new_thread = 'false'
```
I've re-written the documentation block for this (see conversation above) which addresses this.

mikejrizzo commented on Packs/EmailCommunication/Scripts/PreprocessEmail/PreprocessEmail.py:
```diff
@@ -238,45 +254,140 @@ def get_email_related_incident_id(email_related_incident_code, email_original_su
     try:
         email_related_incident_code = email_subject.split('<')[1].split('>')[0]
-        email_original_subject = email_subject.split('<')[-1].split('>')[1]
+        email_original_subject = email_subject.split('<')[-1].split('>')[1].strip()
```
Without the `strip()` the `email_original_subject` would start with a space, since the full subject will be in this format:
<12345678> Email Subject Text

With this change `email_original_subject` will not start with a space.

mikejrizzo commented on Packs/EmailCommunication/Scripts/SendEmailReply/SendEmailReply.py:
```diff
@@ -54,15 +93,174 @@ def send_mail_request(incident_id, email_subject, email_to, reply_body, service_
+        Dict of email thread entries
+    """
+    # Get current email threads from context if any are present
+    incident_context = demisto.executeCommand("getContext", {'id': incident_id})
```
In the added functionality for the "Email Threads" layout, all email messages are stored as list items within the context key `EmailThreads` on the incident they are associated with.  Constructing new emails to send and storing them requires the current entries in `EmailThreads`, which is the purpose of this function.

mikejrizzo commented on Packs/EmailCommunication/Scripts/PreprocessEmail/PreprocessEmail.py:
```diff
@@ -238,45 +254,140 @@ def get_email_related_incident_id(email_related_incident_code, email_original_su
+        }
+        # Add email message to context key
+        try:
+            demisto.executeCommand('executeCommandAt', {
```
I was not able to find any other commands or scripts that would allow for setting context keys on a different incident other than the one being fetched and handled in pre-processing.  In this case a pre-existing incident is found that relates to the incoming email that's being processed, and the details of the new email need to be added as a context key on that related incident. `executeCommandAt` was the only way I could find to modify context on the related incident.

Please let me know if there's a more preferred way to do this that I missed.

mikejrizzo commented on Packs/EmailCommunication/Scripts/SendEmailReply/SendEmailReply.py:
```diff
@@ -222,12 +431,139 @@ def get_mailbox_from_incident_labels(labels):
+    return incidents_details
+
+
+def get_unique_code():
```
For the new "Email Threads" layout and it's added functionality I needed to be able to generate a unique code for first-contact emails being sent from XSOAR since there would not already be one to reference.  I made copies of code from `PreprocessEmail` so that codes are being generated the same way.  I didn't see a way to do it without code re-use, other than perhaps adding these functions to `CommonServerPython`.

mikejrizzo commented on Packs/EmailCommunication/Scripts/DisplayEmailHtmlThread/DisplayEmailHtmlThread.py:
```diff
@@ -0,0 +1,130 @@
+import demistomock as demisto  # noqa: F401
+from CommonServerPython import *  # noqa: F401
```
I have created a set of unit tests for this script.

mikejrizzo commented:
@yuvalbenshalom How would you like to handle updating the documentation on xsoar.pan.dev given that it's stored in a separate repo?
https://xsoar.pan.dev/docs/reference/packs/email-communication

altmannyarden commented:
When the PR is ready, let's do a final meeting with a demo to verify its all working as expected.

yuvalbenshalom commented on Packs/EmailCommunication/Scripts/SendEmailReply/SendEmailReply.py:
```diff
@@ -222,12 +459,436 @@ def get_mailbox_from_incident_labels(labels):
+    return code
+
+
+def reset_fields():
```
please add doctoring that explain the purpose.

mikejrizzo replied:
Done.

yuvalbenshalom commented on Packs/EmailCommunication/Scripts/SendEmailReply/SendEmailReply.py:
```diff
@@ -222,12 +459,436 @@ def get_mailbox_from_incident_labels(labels):
+        return_error(f"Failed to send email via new_thread = 'n/a' branch. Reason: {error}")
+
+
+def multi_thread_new(new_email_subject, new_email_recipients, new_email_body, incident_id, email_codes,
+                     new_email_attachments, files, service_mail, add_cc, add_bcc, mail_sender_instance,
+                     new_attachment_names):
+    """
+            Retrieve all entries in the EmailThreads context key
+    Args:
+        email_codes: The random code that was generated when the incident was created.
+        new_email_body: The email body
+        incident_id: The incident ID.
+        add_cc: The email cc.
+        add_bcc: The email bcc.
+        files: Dictionary of incident file details
+        new_email_subject: The email subject
+        new_email_recipients: The email's recipients
+        service_mail: The service mail (mail listener).
+        new_email_attachments: Files to attach to the new email message
+        mail_sender_instance: The name of the mail sender integration instance
+        new_attachment_names: File names of attachments being sent on the email
+    Returns:
+        String containing result message from send_new_email function
+        """
```
this is the same documentation as in `collect_thread_details` func.
please fix that and add an explanation about the role of this function.

mikejrizzo replied:
Done.

yuvalbenshalom commented on Packs/EmailCommunication/Scripts/SendEmailReply/SendEmailReply.py:
```diff
@@ -222,12 +459,436 @@ def get_mailbox_from_incident_labels(labels):
+def multi_thread_reply(new_email_body, incident_id, email_selected_thread, new_email_attachments, files, add_cc,
+                       add_bcc, service_mail, mail_sender_instance, new_attachment_names):
+    """
+    Retrieve all entries in the EmailThreads context key
```
this is the same description as `collect_thread_details`. please add accurate and informative description.

mikejrizzo replied:
Done.

yuvalbenshalom commented on Packs/EmailCommunication/Scripts/SendEmailReply/SendEmailReply.py:
```diff
@@ -239,25 +900,45 @@ def main():
+    if new_email_attachments:
+        new_attachment_names = ', '.join([attachment.get('name', '') for attachment in new_email_attachments])
+    else:
+        new_attachment_names = 'None'
```
aquestion for later. why are we using here 'None' and not None?
the same as fir the string 'false' instead of False.

mikejrizzo replied:
The `new_attachment_names` variable is a human-readable string that is eventually included in the new `EmailThreads` context entry to store that email.  When the email thread is displayed later in the thread viewer this string is shown so the user can see what file names were present on that particular email message.

yuvalbenshalom reviewed (changes requested):
reviewed the last refactor.
please see my comments.

mikejrizzo commented on Packs/EmailCommunication/.secrets-ignore:
```diff
@@ -1,5 +1,18 @@
+cc_user_1@company.com
+cc_user_2@company.com
+bcc_user@company.com
+BY5PR09ME5460A9F1D8E34A12904AE86EB6199@BY5VR02MB5660.namprd09.prod.outlook.com
+BY5PR03ME5460A9F1D8E34A12904AE86EB6191@BY5VR12MB5660.namprd09.prod.outlook.com
+listName=\"XSOAR
+AAMkAGRcOGZlZTEzLTkyZGDtNGJkNy1iOWMxLYM0NTAwODZhZjlxNABGAAAAAAAP2ksrJ8icRL4Zhadm7iVXBwAkkBJXBb0
+sRJWC0zdXEMqsAAAAAAEMAAAkkBJFBb0fRJWC0zdXEMqsABApcWVYAAA='
+dm7iVXBwAkkBJXBb0sRJWC0zdXEMqsAAAAAAEMAAAkkBJFBb0fRJWC0zdXEMqsABApcWVYAAA=
```
Those are some email message IDs that are included in test data for unit test scripts.  They're not real, but are made to look like actual message IDs so that the test script validates the function works with real-looking data.

mikejrizzo commented on Packs/EmailCommunication/Scripts/PreprocessEmail/PreprocessEmail.py:
```diff
@@ -238,45 +254,140 @@ def get_email_related_incident_id(email_related_incident_code, email_original_su
     return code
 
 
+def create_thread_context(email_code, email_cc, email_bcc, email_text, email_from, email_html,
```
Done.

mikejrizzo commented on Packs/EmailCommunication/Scripts/SendEmailReply/SendEmailReply.py:
```diff
@@ -41,25 +82,210 @@ def execute_reply_mail(incident_id, email_subject, email_to, reply_body, service
+    return incident_email_threads
+
+
+def create_thread_context(email_code, email_cc, email_bcc, email_text, email_from, email_html,
```
Done.

mikejrizzo commented:
@yuvalbenshalom I have submitted a pull request for the documentation update here: https://github.com/demisto/content-docs/pull/1076
